### PR TITLE
merge duplicated linux/windows kube-proxy setup code

### PR DESCRIPTION
**Pull request by danwinship**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
There is currently a lot of duplicated code between the linux (iptables/ipvs) and windows (winkernel) kube-proxy setup. This extracts it out to the shared code, along with a bit of other platform-related refactoring.

extracted from #117436

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/sig network

### CODE REVIEW

**Comment by danwinship** [comment]
Hm... initial push failed CI because replacing `server_others.go` with `server_linux.go` broke the kube-proxy build on OS X, which is surprising given that we don't build kube-proxy on OS X. I "fixed" this, but I feel like this is a build system bug, and I'm wondering if maybe I should just remove the `server_others.go`→`server_linux.go` renaming until it's fixed...

**Comment by dims** [comment]
For a few tools like verify-typecheck.sh it's easier to at least build these things even if we don't ship them

**Comment by BenTheElder** [comment]
> For a few tools like verify-typecheck.sh it's easier to at least build these things even if we don't ship them

IIRC verify-typecheck.sh is setting both GOOS and GOARCH to target the platforms we build for, it shouldn't need to build this for mac.

I have no idea why we'd build kube-proxy for mac, we aren't in releases?

It is nice to have code compile in IDEs and run tests, but if it's calling out to OS specific stuff then that doesn't help a lot.


I would guess we just refactored to add windows support and added `!windows` where `linux` was more approriate.

**Comment by BenTheElder** [comment]
> @BenTheElder don't we have kube-proxy for Mac in downloads (https://kubernetes.io/releases/download/#container-images)? if it was the thing you were mentioning above.

Er no? What would a Mac container image look like?

The kube-proxy image in the release notes is only for Linux, not even Windows.
https://explore.ggcr.dev/?image=registry.k8s.io%2Fkube-proxy%3Av1.27.1

We have Linux and Windows binaries for some node components.
We have Linux only for server components (kube-apiserver etc.)
And for macOS we have kubectl and kubectl plugins AFAIK:

```console
$ gsutil ls gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl-convert
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl-convert.cert
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl-convert.sha256
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl-convert.sha512
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl-convert.sig
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl.cert
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl.sha256
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl.sha512
gs://kubernetes-release/release/v1.27.0/bin/darwin/amd64/kubectl.si
```

**Comment by uablrek** [comment]
I don't understand what's wrong in;
```
kubernetes > ./hack/update-gofmt.sh
go version go1.20.4 linux/amd64
error: size of cmd/kube-proxy/app/server_others.go changed during reading (from 1311 to 1310 bytes)
```

**Comment by danwinship** [comment]
Sorry, I didn't provide a ton of context on the build stuff because I didn't mean for this bug to turn into a discussion of that (see #118020 instead). But to clarify since different people seem to be different parts of the context;

1. `hack/lib/golang.sh` has a bunch of variables that control which things get built on which platforms. On OS X, we intentionally only build `kubectl` and `e2e.test` (more or less). This is not a bug.

2. Prior to this PR, the Windows kube-proxy backend was handled by `cmd/kube-proxy/app/server_windows.go` and the Linux-specific backends (iptables/ipvs) were handled by `cmd/kube-proxy/app/server_others.go`. And in fact, although our normal build processes do not build kube-proxy on OS X, it actually _was_ possible to "`make WHAT=cmd/kube-proxy`" by hand on OS X, and you'd end up with a copy of the Linux kube-proxy which you could run on OS X and it would immediately error out because it couldn't find an "`iptables`" binary.

3. Given that it doesn't make sense to build a Linux-specific kube-proxy backend on non-Linux platforms, I decided to redo the file names and build tags to clarify that the iptables and ipvs kube-proxy backends were only for "`linux`", not for "`others`"/"`!windows`".

4. Surprisingly, given point 1, making the change in point 3 caused verify-typecheck to fail on OS X, because verify-typecheck _was_ trying to build kube-proxy, and kube-proxy now had missing symbols on any platform except `linux` and `windows`.

5. (For now, I fixed point 4 by adding back a stub `server_others.go` file so that now you can build and run kube-proxy on OS X and it will immediately error out with a message about not supporting this platform.)

**Comment by BenTheElder** [comment]
I don't know who is depending on this, but for reference @ 3588d091ff3549b50e9c993b12b169b4caeb105f (current `master`)
```console
$ git rev-parse HEAD
3588d091ff3549b50e9c993b12b169b4caeb105f

$ uname
Darwin

$ go build -o ./kubelet ./cmd/kubelet

$ ./kubelet
E0516 11:13:54.187405   69372 run.go:74] "command failed" err="failed to construct kubelet dependencies: mkdir /var/lib/kubelet: permission denied"

$ go build -o ./kube-proxy ./cmd/kube-proxy

$ ./kube-proxy
I0516 11:17:04.538852   69859 server.go:233] "Warning, all flags other than --config, --write-config-to, and --cleanup are deprecated, please begin using a config file ASAP"
I0516 11:17:04.539004   69859 server_others.go:75] "Using iptables proxy"
I0516 11:17:04.539251   69859 server.go:531] "Neither kubeconfig file nor master URL was specified, falling back to in-cluster config"
E0516 11:17:04.539272   69859 server.go:484] "Error running ProxyServer" err="unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"
E0516 11:17:04.539290   69859 run.go:74] "command failed" err="unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"
```

**Comment by danwinship** [comment]
> I "fixed" this, but I feel like this is a build system bug, and I'm wondering if maybe I should just remove the `server_others.go`→`server_linux.go` renaming until it's fixed...

(done)

**Comment by aojea** [comment]
> Unexpected final error while getting *v1.Pod: etcdserver: request timed out
> At one point, however, the function did return successfully.

weird failures 

/test pull-kubernetes-e2e-kind-ipv6

File: cmd/kube-proxy/app/server_others.go [hunk 1]
```diff
@@ -329,7 +328,6 @@ func newProxyServer(
 		if err != nil {
 			return nil, fmt.Errorf("unable to create proxier: %v", err)
 		}
-		proxymetrics.RegisterMetrics()
```
Comment (aojea) [inline cmd/kube-proxy/app/server_others.go:332]: this was kind of weird, no? initializing in two different places instead doing just in one, maybe a copy paste thing

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -319,7 +322,7 @@ func (o *Options) Run() error {
 		return cleanupAndExit()
 	}
 
-	proxyServer, err := NewProxyServer(o)
```
Comment (aojea) [inline cmd/kube-proxy/app/server.go:322]: I hope nobody is importing this

File: cmd/kube-proxy/app/server.go [hunk 3]
```diff
@@ -514,13 +517,63 @@ type ProxyServer struct {
 	Client        clientset.Interface
 	Broadcaster   events.EventBroadcaster
 	Recorder      events.EventRecorder
-	Conntracker   Conntracker // if nil, ignored
 	NodeRef       *v1.ObjectReference
 	HealthzServer healthcheck.ProxierHealthUpdater
+	Hostname      string
```
Comment (aojea) [inline cmd/kube-proxy/app/server.go:522]: this could be internal variables

Comment (danwinship) [reply cmd/kube-proxy/app/server.go:522]: you mean `hostname` rather than `Hostname`? yeah, the whole struct is wrong that way

Comment (aojea) [reply cmd/kube-proxy/app/server.go:522]: yep yep

**Pull request merged**